add function `Sys.username()`

barucden opened the pull request:
The commit introduces a new function to Base which returns the current user's username retrieved from the password database.

Resolves #48302
Closes #48928

barucden committed:
Base: add function `username` to `Base.Sys`

The commit introduces a new function to `Base.Sys` which returns the
current user's username retrieved from the password database.

Resolves #48302
Closes #48928

Co-authored-by: Steven G. Johnson <stevenj@mit.edu>

barucden commented on base/env.jl:
```diff
@@ -244,3 +244,30 @@ function withenv(f, keyvals::Pair{T}...) where T<:AbstractString
+    username = Filesystem.getusername(uid)
+    isnothing(username) || return username
+    # the user was not found or the username was empty
+    throw(ErrorException("username not found in the password database"))
```
Is there a more suitable exception type?

vtjnash replied:
The system usually reports ENOENT for it as an IOError

JeffFessler commented:
Looks terrific!  Thanks for taking the lead on this.

PallHaraldsson commented:
>In case of a name conflict, one needs to prefix the reference (Base.symbol) to get the Base definition.

Continuing discussion here. I think you're actually right, but I'm thinking also, would username() you are defining be likely used in Packages (and assuming they do unprefixed, since they think ok, because it's exported)? Let's assume that. I'm not sure what happens if you import a package, and define your own username(), maybe you did that long ago, now adding a dependency, will you redefine username() that the package is using to use your definition? I'm not sure it will happen, but if, then this is a breaking change (or a potential footgun for users).

If this is possible, then there are also other changes that might help. Packages are precompiled, and I'm not sure I want it possible that they get changed under your feet. If you use e.g. Python, then I think you can redefine pretty much anything, as in Julia, it's an even more dynamic language, but if you're calling precompiled packages, as often done for speed, then it's just not technically possible to change them, you could need a C compiler. Possibly that can and should happen with Julia precompiled packags, that you can't redefine (also would get rid of invalidations?).

JeffFessler commented:
I don't know the answers to @PallHaraldsson' questions precisely, but the general concern about name conflicts would seem to apply to proposals to add _any_ new functions to Julia.  Yet additions are being made; for example `stack` was added in 1.9:  https://docs.julialang.org/en/v1/base/arrays/#Base.stack
The name `stack` is just as generic as `username` so it is likely to have existed in some other packages/code.

stevengj commented:
We don't consider exporting new functions to be breaking changes IIRC.

stevengj commented on base/env.jl:
```diff
@@ -244,3 +244,30 @@ function withenv(f, keyvals::Pair{T}...) where T<:AbstractString
+
+!!! note
+    The username is determined via the user account and password database
+    that is available only on Unix-like platforms.
```
Why?  On Windows we can call [`GetUserNameExA`](https://learn.microsoft.com/en-us/windows/win32/api/secext/nf-secext-getusernameexa) or similar, no?

Doesn't libuv do this for us automatically?

stevengj replied:
According to the [libuv source code](https://github.com/libuv/libuv/blob/da527d8d2a908b824def74382761566371439003/src/win/util.c#L1262), its `getpwuid` function calls `GetUserNameW` on Windows, so it seems like it should work.

barucden replied:
I saw this https://github.com/JuliaLang/julia/blob/ba984670cf11721928ec8eb528ce2052a2f7b7e2/src/sys.c#L224 and gave up on Windows.

Julia calls `uv_os_get_passwd2` (https://github.com/JuliaLang/julia/blob/master/base/libc.jl#L498) which does not seem to be implemented for Windows (https://github.com/libuv/libuv/blob/da527d8d2a908b824def74382761566371439003/src/win/util.c#L1301)?

stevengj replied:
You could always call the [`GetUserNameW`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getusernamew) function directly, instead of going through `getuid`.

Something like:
```jl
buf = Vector{Cwchar_t}(undef, 257) # UNLEN+1
len = Ref(UInt32(257))
success = ccall(:GetUserNameW, stdcall, UInt32,
                (Ptr{Cwchar_t}, Ref{UInt32}), buf, len)
windowserror("GetUserNameW", success == 0)
return transcode(String, resize!(buf, len[]))
```

stevengj replied:
Or simply call `uv_os_get_passwd`, rather than `uv_os_get_passwd2`, since it looks like the former *is* supported on Windows.

barucden replied:
Thank you, Steven. Can you check it now? What do you think about defining the function in `libc.jl`?

stevengj commented on test/env.jl:
```diff
@@ -181,3 +181,12 @@ end
+        passwd = Libc.getpwuid(Libc.getuid())
+        @test username() == passwd.username
+    else
+        @test !isempty(username())
```
This should no longer be empty on Windows, no?

stevengj commented on test/env.jl:
```diff
@@ -181,3 +181,12 @@ end
+        @test username() == passwd.username
+    else
+        @test !isempty(username())
+    end
```
Isn't there an environment variable one could compare to on Windows?

barucden replied:
Apparently, `USERNAME` should work. Although I am not sure if it always coincides with that returned by `GetUsername`.

barucden commented:
The Windows CI finished successfully. The failing tests are unrelated.

PallHaraldsson commented:
Since you actually added to Libc, should maybe NOT export from Base, since not needed, i.e. Libc.username() would work, and this wouldn't be used much anyway?

I'm thinking exporting from Base could always be added later, and I still have concerns about it:

>We don't consider exporting new functions to be breaking changes IIRC.

That is perfectly fine to do from e.g. Libc or any non-Base module. Are you sure there are no practical implications for Base, and also that thought ok. I seem to recall a doc PR suggesting changes to regarding `using` in practice i.e. proposing NOT exporting from packages. It seems like the same issue, I just haven't tracked that PR down.

>the general concern about name conflicts would seem to apply to proposals to add any new functions to Julia.

No, but yes, from Base, since using it is implicit. We get away with it if functions aren't conflicting in practice.

As annoying as fully qualifying is, it's the norm in e.g. Python, and after namspaces in C++, for some reason e.g. cout that used to work needs to be std::cout, so I think they have thought this through, that it might be better.

barucden commented:
> Since you actually added to Libc

I put it to the `Libc` submodule because that's where the `Cpasswd` structure is defined.

> should maybe NOT export from Base,

I don't know. If `username` is supposed to be a part of the user-facing API, then I'd say it should be exported. If it is not supposed to be a part of the user-facing API, then there is little reason to add it at all because it is not used anywhere in Base.

Anyway, I do not feel qualified to be the judge of that. I'll update the PR to whatever the maintainers suggest.

PallHaraldsson commented:
>If username is supposed to be a part of the user-facing API

It still would be, we now have the new public keyword for that, to make it clear. It would be documented, and if you worry about it not being seen or overlooked in the docs, then yes, I do not find ?malloc with help, unless I do ?Libc.malloc, but maybe that should be changed. I think the help should find all public, also from Libc or stdlibs not yet imported. The help text should then add, "In Libc, so needs to be prefixed with Libc."

It's not possible to unexport later, or it would be a breaking change, but adding an export from Base could always be added.

Do you think this function would be frequently used? And do you think it would never be a problem/conflict with exporting (even for packages)?

stevengj commented:
`Libc.username` seems like an odd place for this, since Libc exports generally correspond directly to low-level libc functions, and this is not such a function.

barucden commented:
`Libc` also defines `getpid` or `gethostname` though. But I don't mind moving `username` elsewhere. Do you have any suggestion?

PallHaraldsson commented:
Those are actually defined in C standard library(?), and Libc (and exported from Base), possibly `getppid` should be added there too.

A better precident as exceptions would be FormatError and GetLastErro (only defined in Windows):

https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror

I suggested in Libc, since to get the username, as a string, it seemed to actually need to use Libc, to get user id, so at least related, if I recall. I admit it may be less discoverable. I would also be ok with unexported, but marked public, as Base.getusername. It's best if we chose wisely which prefix to not need to allow both... I'm assuming stuff in Base needs not be in the sysimage... I don't like it bloated for startup cost reasons, not just for name-space pollution.

stevengj commented:
I think we should just export it from Base, analogous to `homedir()`.   We've allowed new exports in Julia minor versions many times already.

JeffFessler commented:
I'll upvote `Sys.username` and I am fine with no export.
Right now `Sys` is a funny collection of very low-level stuff like `cpu_info` and very high level stuff like `isunix`
https://github.com/JuliaLang/julia/blob/master/base/sysinfo.jl
so it might not be where people would look first for `username` but as long as it is documented it can be found there.

(In my dreams I would type `user<tab>` at the REPL and it would offer me the option of `Sys.username`.)

barucden commented:
Alright, I moved the user-facing function to `Sys`. `Libc` now defines `getpw` which is the lower-level function.

PallHaraldsson commented:
I'm ok with merging as is since this does what the title says (after it's changed from Base: -> Sys:)

I saw this was marked as a "good first issue". I hope I didn't make it too hard on you... :)

One question though on the new getpw, though, should it be exported like others, e.g. gethostname? And I' not sure it makes a difference, since it will be used as Libc.getpw. So what is the exporting for there, all under Libc, need qualifying already?! If it's just to signal part of the API, then the new public keyword is for that, and could alternatively be used.

I see it uses a subset of the full struct, and maybe that's an argument for keeping NOT part of the API?

barucden commented:
> I saw this was marked as a "good first issue". I hope I didn't make it too hard on you... :)

All good. It is not my first PR too.

> One question though on the new getpw, though, should it be exported like others, e.g. gethostname?

I do not think it should be exported. It is almost the same as `getpwuid`, which is not exported.

stevengj commented on base/sysinfo.jl:
```diff
@@ -567,4 +568,25 @@ function which(program_name::String)
+"""
+function username()
+    pw = Libc.getpw()
+    isempty(pw.username) || return pw.username
+    throw(ErrorException("username was empty"))
```
I think it's a bit clearer to put the `return` statement last:
```suggestion
    isempty(pw.username) && throw(ErrorException("empty username"))
    return pw.username
```

stevengj commented on base/libc.jl:
```diff
@@ -493,6 +493,26 @@ struct Group
+function getpw()
+    ref_pd = Ref(Cpasswd())
+    ret = ccall(:uv_os_get_passwd, Cint, (Ref{Cpasswd},), ref_pd)
+    ret != 0 && Base.uv_error("getpw", ret)
```
```suggestion
    Base.uv_error("getpw", ret)
```
The `ret != 0` check is unnecessary because `uv_error` only throws if `ret < 0`.   (Similarly in `getpwuid`.)

barucden replied:
I think it is necessary in `getpwuid` because we just want to return `nothing` if `throw_error=false`.

stevengj commented on base/libc.jl:
```diff
@@ -493,6 +493,26 @@ struct Group
     mem::Vector{String}
 end
 
+# Gets a subset of the password file entry
```
```suggestion
# Gets password-file entry for default user, or a subset thereof
```

stevengj commented on NEWS.md:
```diff
@@ -50,6 +50,7 @@ New library functions
 * `diskstat(path=pwd())` can be used to return statistics about the disk. ([#42248])
 * `copyuntil(out, io, delim)` and `copyline(out, io)` copy data into an `out::IO` stream ([#48273]).
 * `eachrsplit(string, pattern)` iterates split substrings right to left.
+* `Sys.username()` can be used to return the current user's username. ([#51897])
```
```suggestion
* `Sys.username()` can be used to return the current user's username ([#51897]).
```

barucden commented:
Thank you. I squashed all the commits, updated the docs (`username` -> `Sys.username`), and rebased to master.

The PR should be ready for merging after CI finishes.

vtjnash commented on base/sysinfo.jl:
````diff
@@ -567,4 +568,25 @@ function which(program_name::String)
+To retrieve a username that is overridable via an environment variable,
+e.g., `USER`, consider using
+```julia
+user = get(ENV, "USER", Sys.username())
````
```suggestion
user = get(Sys.username, ENV, "USER")
```

barucden committed:
Update base/sysinfo.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

barucden commented on base/sysinfo.jl:
```diff
@@ -567,4 +568,25 @@ function which(program_name::String)
+"""
+function username()
+    pw = Libc.getpw()
+    isempty(pw.username) && throw(ErrorException("empty username"))
```
```suggestion
    isempty(pw.username) && throw(Base.IOError("empty username", Base.UV_ENOENT))
```
So, this is preferred, @vtjnash?

vtjnash replied:
Maybe like this:
```suggestion
    isempty(pw.username) && Base.uv_error("username", Base.UV_ENOENT)
```
to match this behavior:
```
julia> Libc.getpwuid(typemax(UInt))
ERROR: IOError: getpwuid: no such file or directory (ENOENT)
```

barucden committed:
Update base/sysinfo.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
